src/classes/vehicle.py

```diff
@@ -87,73 +90,121 @@ def update_cmd(old_cmd: Command, t: np.array, a: np.array, elapsed_time: float=0
-                continue
-            elif distance == safety_distance:
-                required_deceleration = 0
-            else:
-                # required deceleration to stop car at fake distance away
-                required_deceleration = (final_velocity**2 - initial_velocity**2) / (2 * (distance - safety_distance))
-            
-            new_t = np.array([cur_time, cur_time + 0.1])
-            new_a = np.array([required_deceleration, leading_vehicle.acceleration])
-            
-            vehicle.command = update_cmd(vehicle.command, new_t, new_a, cur_time)
+        check_traffic_lights(vehicle, cur_time)
+
+
+def handle_leading_vehicle(vehicle, cur_time: float) -> None:
+    """Handle the case where there is a leading vehicle."""
     
-        elif abs(vehicle.velocity - vehicle.default_velocity) > 0.01:
-            acceleration_distance = 10
-            required_deceleration = (vehicle.default_velocity**2 - initial_velocity**2) / (2 * acceleration_distance)
-            
-            new_t = np.array([cur_time, cur_time + 0.1])
-            new_a = np.array([required_deceleration, vehicle.acceleration])   
-            
+    # we want to stop the car "Safety Distance" away from leading car
+    safety_distance = 6
+    emergency_distance = 5
+
+    leading_vehicle = vehicle.leading_vehicle
+    initial_velocity = vehicle.velocity
+    final_velocity = leading_vehicle.velocity
+    distance = abs(vehicle.route_position - leading_vehicle.route_position)
+
+    # this prevents division of negative and 0
+    if distance < emergency_distance:
+        vehicle.velocity = 0
+        print("Invalid command: Distance to leading vehicle is too short")
+        return
+    
+    # required deceleration to stop car at fake distance away
+    required_deceleration = calculate_deceleration(final_velocity, initial_velocity, distance, safety_distance)
+
+    new_t = np.array([cur_time, cur_time + 0.01])
+    new_a = np.array([required_deceleration, leading_vehicle.acceleration])
+    vehicle.command = update_cmd(vehicle.command, new_t, new_a, cur_time)
+
+
+def handle_no_leading_vehicle(vehicle, cur_time: float) -> None:
+    """Handle the case where there is no leading vehicle."""
+    acceleration_distance = 10
+    initial_velocity = vehicle.velocity
+
+    if abs(vehicle.velocity - vehicle.default_velocity) > 0.01:
+        required_acceleration = (vehicle.default_velocity**2 - initial_velocity**2) / (2 * acceleration_distance)
+
+        new_t = np.array([cur_time, cur_time + 0.01])
+        new_a = np.array([required_acceleration, vehicle.acceleration])
+        vehicle.command = update_cmd(vehicle.command, new_t, new_a, cur_time)
+
+
+def check_traffic_lights(vehicle, cur_time: float) -> None:
+    """Check traffic lights and update commands based on their states."""
+    for r in vehicle.route.pos_to_edge_map:
+        if vehicle.route_position < r[1]:
```

Comment:
we need vehicles to respond to traffic lights no matter what edge they're on - this will not work right?

Reply:
i'm not sure what you mean? could you give me more detail of what you're thinking?